### PR TITLE
Fix and improve relative binning in HeterodynedTransientLikelihoodFD

### LINKED ISSUE #108: Check and improve relative binning
The current relative binning (RB) code in Jim now is not exactly the same as the one in Bilby. Checks are needed in terms of the following.

- What are the differences?
- Just different binning scheme? (Jim: fixed `n_bins`)
- How to set up Jim/Bilby to get the exact same result

Also, if the version from Bilby is "better", we should improve the version in Jim by porting stuff from Bilby.

### PR DESCRIPTION
**Pull request by thomasckng**
Closes #108.

## Summary

This PR corrects three bugs in `HeterodynedTransientLikelihoodFD` and aligns the implementation with Bilby's relative-binning reference (arXiv:1806.08792).

---

### 1. Correct gamma set in `_max_phase_diff`

The old implementation used `gamma = arange(-5, 6) / 3` (11 terms, including zero) with a `f_star`-based normalisation that is not equivalent to Bilby's. Replaced with Bilby's 5 physically-motivated PN/IMR terms:

$$\gamma \in \{-5/3,\ -2/3,\ 1,\ 5/3,\ 7/3\}$$

Each term is normalised by `d_alpha` so that its contribution spans exactly `chi * 2π` rad across `[f_low, f_high]`. The array is offset so it starts at zero (cumulative phase from `f_low`).

### 2. Correct r0/r1 estimators in `_likelihood`

The old code estimated the ratio polynomial using a center–left off-centre finite difference:
```python
r0 = waveform_center / waveform_center_ref
r1 = (waveform_low / waveform_low_ref - r0) / (f_low - f_center)
```

Replaced with Bilby's bin-edge average:
```python
r0 = (r_low + r_high) / 2
r1 = (r_high - r_low) / bin_width
```

This requires evaluating the reference and proposal waveforms at both bin edges. `waveform_high_ref` (evaluated at `freq_grid_high`) replaces `waveform_center_ref`, and `bin_widths` is stored as an instance attribute.

### 3. Adaptive epsilon-based binning

Added an `epsilon` parameter — the maximum allowed phase change per bin in radians — as the primary way to control bin count, matching Bilby's interface:

```python
n_bins = max(1, int(total_phase / epsilon))
```

- `epsilon=0.5` (the new default when neither argument is given) yields ~60 bins for a typical CBC signal with `f=[20, 512]` Hz.
- `n_bins` is still accepted but is now **mutually exclusive** with `epsilon`; passing both raises `ValueError`.
- Non-positive values of either parameter raise `ValueError`.
- The CLI (`CLIHeterodynedConfig`) gains a corresponding `epsilon` field; both default to `None`.

### 4. Fix NaN at waveform boundary

The masking condition used `freq_grid_center <= f_waveform_max`, which could include bins whose upper edge (`freq_grid_high`) extended beyond the waveform's valid range. At those frequencies the reference waveform is zero, causing `r_high = 0 / 0 = NaN`. Fixed by using `freq_grid_high <= f_waveform_max` as the upper-bound condition.

### 5. Rename helper methods to private

`max_phase_diff`, `make_binning_scheme`, and `compute_coefficients` are internal helpers not intended as public API. Renamed to `_max_phase_diff`, `_make_binning_scheme`, and `_compute_coefficients`. `_make_binning_scheme` no longer accepts `epsilon`; the resolution from epsilon to integer `n_bins` is done in `__init__` before calling it.

### 6. CLI and docs

- `CLIHeterodynedConfig` gains `epsilon: Optional[float] = None` alongside `n_bins: Optional[int] = None`.
- `build_likelihood` passes `epsilon` to the constructor and now logs `likelihood.n_bins` (the actual post-construction bin count) rather than the config value.
- `docs/guides/cli.md` updated to reflect the new epsilon-based TOML configuration.
- All docstrings updated.

---

## Tests

- Unit tests updated: `waveform_center_ref` → `waveform_high_ref`; `freq_grid_high` attribute assertion added. All 13 tests pass.
- Cross-validation tests updated: use `epsilon=0.5` directly instead of matching `bilby_likelihood.number_of_bins`.

## Release Notes

* **New Features**
  * Added `epsilon`-based heterodyne/relative-binning configuration, alongside optional fixed-bin control via `n_bins`.

* **Changes**
  * Heterodyne binning is now controlled by at most one of `epsilon` or `n_bins`; if neither is set, `epsilon=0.5` is used.
  * Likelihood marginalisation inputs now accept `{…}` dictionaries and `True` shorthand (empty `{}` disables the mode).

* **Documentation**
  * Updated the CLI guide for `[likelihood.heterodyne]` to reflect the new binning options.

* **Tests**
  * Adjusted cross-validation and updated heterodyne initialisation/shape assertions for the new scheme.

### CODE REVIEW

**Comment by thomasckng** [comment]
@coderabbitai review

File: src/jimgw/core/single_event/likelihood.py [hunk 1]
```diff
@@ -795,38 +802,52 @@ def _likelihood(self, params: dict[str, Float]) -> FloatScalar:
...
+        See also Eq.(7) in arXiv:2302.05333.
         """
-        gamma = jnp.arange(-5, 6) / 3.0
+        gamma = jnp.array([-5 / 3, -2 / 3, 1.0, 5 / 3, 7 / 3])
         freq_2D = jax.lax.broadcast_in_dim(freqs, (freqs.size, gamma.size), [0])
         f_star = jnp.where(gamma >= 0, f_high, f_low)
         summand = (freq_2D / f_star) ** gamma * jnp.sign(gamma)
-        return 2 * jnp.pi * chi * jnp.sum(summand, axis=1)
+        dphi = 2 * jnp.pi * chi * jnp.sum(summand, axis=1)
+        return dphi - dphi[0]
```
[the remark this answers is not in the document]
Comment (SSL32081) [reply src/jimgw/core/single_event/likelihood.py:821]: The gamma terms have been normalised in this implementation, and the explicit reference to `d_alpha` has been removed from the documentation.

**Review by SSL32081 — APPROVED** [review]
The biggest changes in this PR are the new `gamma` list, the minimal phase resolution `epsilon` and the slightly modified waveform ratios.
* As mentioned in #123, the list of `gamma` should have an impact mostly in the high-frequency regime. In fact, if instead of replacing the existing list entirely, we simply add the 7/3 order, it might achieve the same result and perhaps outperform Bilby at the potential cost of additional memory.
* The existing implementation of the waveform ratio is, in principle, the same as the new one, especially when the frequency bins are small. This change should bring minimal impact to the overall results.

File: src/jimgw/core/single_event/likelihood.py [hunk 2]
```diff
@@ -1322,7 +1322,7 @@ def _compute_tau_dtaudf(self, f: Float) -> tuple[Float, Float]:
 
     def _find_starting_frequency(
         self, duration: float, f_now: float
-    ) -> tuple[Optional[float], Optional[float]]:
+    ) -> tuple[Float, Float]:
```
Comment (SSL32081) [inline src/jimgw/core/single_event/likelihood.py]: But I thought in some cases, it could return with `None, None`?

Comment (thomasckng) [reply src/jimgw/core/single_event/likelihood.py]: Ya, my bad. Fixing

Comment (thomasckng) [reply src/jimgw/core/single_event/likelihood.py]: Pre-commit doesn't pass with `Optional`

Comment (SSL32081) [reply src/jimgw/core/single_event/likelihood.py]: I am guessing it was the `jnp.sqrt(...)` instead, which would return with an array? I am not sure.
Coz I see it says it can't cast an array to `Float`?

Comment (thomasckng) [reply src/jimgw/core/single_event/likelihood.py]: Right, changed to `Float` instead of `float`

**Comment by thomasckng** [comment]
@SSL32081 Should line 749-752 just be A0, A1, ..., so without the `[mask_heterodyne_center]` index?

**Comment by thomasckng** [comment]
> @SSL32081 Should line 749-752 just be A0, A1, ..., so without the `[mask_heterodyne_center]` index?

I think `_compute_coefficients(...)` already handled that with the `start_idx` stuff.

**Comment by SSL32081** [comment]
> @SSL32081 Should line 749-752 just be A0, A1, ..., so without the `[mask_heterodyne_center]` index?

I have seen that coderabbit comment, I don't think so, coz
1. In principle, the mask has a different shape than `freq_grid_center`, if what it says were true, we should have met an error whenever the `mask_heterodyne_center` is non-trivial. 
2. If I am reading the `_compute_coefficients` function correctly, I think it is computed on `detector.sliced_frequencies` × `freq_grid`, and then summed over `freq_grid`, so the result should be the same size of `detector.sliced_frequencies`, which should be the same as the mask.

**Comment by SSL32081** [comment]
> > @SSL32081 Should line 749-752 just be A0, A1, ..., so without the `[mask_heterodyne_center]` index?
> 
> I think `_compute_coefficients(...)` already handled that with the `start_idx` stuff.

I think in the `_compute_coefficients` function, there are three frequencies, 
* `detector.sliced_frequencies` → size A
* `freq_grid` → size B
* `freq_grid_center` → size B

and A ≥ B. 

Before the summation at the end, I think the `freq_shift_matrix` has shape `A × B` and then summed over `axis = 1`, which left with a size A array.

**Comment by thomasckng** [comment]
@SSL32081 Does the following test make sense to you?

```
    def test_coefficients_are_not_reindexed_after_frequency_trimming(
        self, detectors_and_waveform, monkeypatch
    ):
        ifos, waveform, fmin, fmax, gps = detectors_and_waveform

        def fake_compute_coefficients(data, h_ref, psd, freqs, f_bins, f_bins_center):
            coeffs = jnp.arange(len(f_bins_center), dtype=jnp.float64)
            return coeffs, coeffs + 100, coeffs + 200, coeffs + 300

        monkeypatch.setattr(
            HeterodynedTransientLikelihoodFD,
            "_compute_coefficients",
            staticmethod(fake_compute_coefficients),
        )

        def reference_waveform(frequencies, params):
            waveform_sky = waveform(frequencies, params)
            mask = frequencies >= 80.0
            return {
                polarization: jnp.where(mask, strain, jnp.zeros_like(strain))
                for polarization, strain in waveform_sky.items()
            }

        likelihood = HeterodynedTransientLikelihoodFD(
            detectors=ifos,
            waveform=waveform,
            reference_waveform=reference_waveform,
            f_min=fmin,
            f_max=fmax,
            trigger_time=gps,
            n_bins=32,
            reference_parameters=example_params(),
        )
        assert likelihood.freq_grid_low[0] > fmin

        expected = jnp.arange(likelihood.n_bins, dtype=jnp.float64)
        for detector in ifos:
            assert jnp.array_equal(likelihood.A0_array[detector.name], expected)
            assert jnp.array_equal(likelihood.A1_array[detector.name], expected + 100)
            assert jnp.array_equal(likelihood.B0_array[detector.name], expected + 200)
            assert jnp.array_equal(likelihood.B1_array[detector.name], expected + 300)
```

**Comment by thomasckng** [comment]
> > @SSL32081 Should line 749-752 just be A0, A1, ..., so without the `[mask_heterodyne_center]` index?
> 
> I have seen that coderabbit comment, I don't think so, coz
> 
> 1. In principle, the mask has a different shape than `freq_grid_center`, if what it says were true, we should have met an error whenever the `mask_heterodyne_center` is non-trivial.
> 2. If I am reading the `_compute_coefficients` function correctly, I think it is computed on `detector.sliced_frequencies` × `freq_grid`, and then summed over `freq_grid`, so the result should be the same size of `detector.sliced_frequencies`, which should be the same as the mask.

Seems like JAX clamp the out-of-bounds indices. Not really sure. I am thinking of a test case to test whether we are right

**Comment by SSL32081** [comment]
(I didn't know you could update the class in-place like so, WOW

If you can do that, can you have the following in the test instead:

```python
def fake_compute_coefficients(self, data, h_ref, psd, freqs, f_bins, f_bins_center):
    coeffs = jnp.arange(len(f_bins_center), dtype=jnp.float64)
    assert len(freqs) >= len(f_bins)
    assert len(f_bins) == len(f_bins_center)
    assert len(self.n_bins) == len(freqs)
    return coeffs, coeffs + 100, coeffs + 200, coeffs + 300
```

make it into an instance method too (not just staticmethod).

If you can somehow guarantee `freqs` and `f_bins` have different sizes too, that would be great, which I think is what your `reference_waveform` is doing right?

I added the `    assert len(self.n_bins) == len(freqs)` because I see you are comparing the array with the `n_bins` at the end, which I am not 1000% sure it is the same size as `freqs`, which is the key thing to test.

**Comment by thomasckng** [comment]
Something like the new commit? This doesn't pass now if I don't remove the `[mask_heterodyne_center]`s.

**Comment by SSL32081** [comment]
> Something like the new commit? This doesn't pass now if I don't remove the `[mask_heterodyne_center]`s.

I see, gimme a couple minutes, let me check it locally, I want to see the sizes of the arrays. Coz it sounds a bit awkward to me, I want to see which part of it went wrong.

**Comment by thomasckng** [comment]
@SSL32081 Final check then I will merge
